pandas/core/dtypes/cast.py

```diff
@@ -1926,7 +1926,12 @@ def np_can_hold_element(dtype: np.dtype, element: Any) -> Any:
                         # i.e. there are pd.NA elements
                         raise LossySetitemError
                 return element
+            # GH 57338 check boolean array set as object type
+            if tipo.kind == "O" and isinstance(element, np.ndarray):
```

Comment:
Could you add a test where `element` has a `np.nan` value too?

Reply:
I added in `test_can_hold_element.py`.